website/docs/validation/forbidden-sysctls.md

```diff
@@ -51,65 +51,86 @@ spec:
                 type: string
   targets:
     - target: admission.k8s.gatekeeper.sh
-      rego: |
-        package k8spspforbiddensysctls
-
-        import data.lib.exclude_update.is_update
-
-        # Block if forbidden
-        violation[{"msg": msg, "details": {}}] {
-            # spec.securityContext.sysctls field is immutable.
-            not is_update(input.review)
-
-            sysctl := input.review.object.spec.securityContext.sysctls[_].name
-            forbidden_sysctl(sysctl)
-            msg := sprintf("The sysctl %v is not allowed, pod: %v. Forbidden sysctls: %v", [sysctl, input.review.object.metadata.name, input.parameters.forbiddenSysctls])
-        }
-
-        # Block if not explicitly allowed
-        violation[{"msg": msg, "details": {}}] {
-            not is_update(input.review)
-            sysctl := input.review.object.spec.securityContext.sysctls[_].name
-            not allowed_sysctl(sysctl)
-            msg := sprintf("The sysctl %v is not explicitly allowed, pod: %v. Allowed sysctls: %v", [sysctl, input.review.object.metadata.name, input.parameters.allowedSysctls])
-        }
-
-        # * may be used to forbid all sysctls
-        forbidden_sysctl(_) {
-            input.parameters.forbiddenSysctls[_] == "*"
-        }
-
-        forbidden_sysctl(sysctl) {
-            input.parameters.forbiddenSysctls[_] == sysctl
-        }
-
-        forbidden_sysctl(sysctl) {
-            forbidden := input.parameters.forbiddenSysctls[_]
-            endswith(forbidden, "*")
-            startswith(sysctl, trim_suffix(forbidden, "*"))
-        }
-
-        # * may be used to allow all sysctls
-        allowed_sysctl(_) {
-            input.parameters.allowedSysctls[_] == "*"
-        }
-
-        allowed_sysctl(sysctl) {
-            input.parameters.allowedSysctls[_] == sysctl
-        }
-
-        allowed_sysctl(sysctl) {
-            allowed := input.parameters.allowedSysctls[_]
-            endswith(allowed, "*")
-            startswith(sysctl, trim_suffix(allowed, "*"))
-        }
-      libs:
-        - |
-          package lib.exclude_update
-
-          is_update(review) {
-              review.operation == "UPDATE"
-          }
+      code:
+      - engine: K8sNativeValidation
+        source:
+          variables:
+          - name: sysctls
+            expression: '!has(object.spec.securityContext) ? [] : !has(object.spec.securityContext.sysctls) ? [] : object.spec.securityContext.sysctls'
+          - name: forbiddenSysctls
+            expression: |
+              !has(variables.params.forbiddenSysctls) ? [] :
+              variables.sysctls.filter(sysctl,
+                  variables.params.forbiddenSysctls.exists(forbiddenSysctl, sysctl.name == forbiddenSysctl || (forbiddenSysctl.endsWith("*") && string(sysctl.name).matches("^" + string(forbiddenSysctl).replace("*", ".*") + "$"))))
```

Comment:
I don't think we can use regex here, sadly, since sysctls use `.` (which is a regex command char) in their allowed character set:

https://kubernetes.io/docs/tasks/administer-cluster/sysctl-cluster/#safe-and-unsafe-sysctls

Since we only allow prefix matching, I suggest the following (context links at bottom):

- create variables called `allowedSysctlPrefixes` and `disallowedSysctlPrefixes` (or similar) and a companion variable called `allowedSysctlExplicit` and `disallowedSysctlExplicit` (or similar).
- Iterate through the constraint parameters, finding entries that end with `*`, adding them to the glob list for their respective parameters (removing the trailing `*`)
- We'll also need to special-case the check for `*` if it exists, maybe set a boolean variable called `allAllowed` and/or `allDisallowed`? Missing parameter should yield `allAllowed` == true and `allDisallowed` == false for allowedSysctls/forbiddenSysctls respectively
- create still another variable `violatingSysctls` or similar iterate through `spec.securityContext.sysctls`, running the following checks (with this precedence):
  - If `allDisallowed`, add the sysctl name to `violatingSysctls`
  - If `disallowedSysctlPrefixes` satisfies a `sysctl.startsWith(prefix)` check, add the sysctl name to violatingSysctls
  - If `disallowedSysctlExplicit` contains the sysctl (`in` keyword), add the sysctl name to `violatingSysctls`
  - if allowed check fails, add the sysctl name to `violatingSysctls` allowed check performs (in order of precedence)
    - If `allAllowed`, check succeeds
    - If `allowedSysctlPrefixes` satisfies a `sysctl.startsWith(prefix)` check, check succeeds
    - If `allowedSysctlExplicit` containers the sysctl (`in` keyword), check succeeds
    - else check fails

When I say "add X to the list if condition is met", that's probably best done using the `list.filter()` macro. Performing mutations, like stripping the `*` character can be done using the `list.map()` macro and use of the extended strings library.

For a conservative string replacement (only removing the very last `*`, in case there is a `*` mid-string:

- Reverse the string
- Call `string.Replace("*", "", 1)`
- Reverse the string again

Another option could be to use `strings.Split()` , pop off the last entry from the list using `lists.Slice()`, then use `strings.join()`

Alternately, if we know `*` cannot be a valid character for a sysctl (which seems likely) we can just call string.Replace() directly.

I mention "extended CEL library".

K8s has its own CEL environments, defined here:

https://github.com/kubernetes/apiserver/blob/v0.30.0/pkg/cel/environment/base.go

Which defines some k8s-specific functions. It also incorporates extensions from cel-go, defined here:

https://github.com/google/cel-go/blob/master/ext

Reply:
Also adding link to CEL macros, for completeness:

https://github.com/google/cel-spec/blob/v0.6.0/doc/langdef.md#macros

Reply:
I started with `allDisallowed` but after more testing, I dont think we need it. PTAL